lib/api/src/index.tsx

```diff
@@ -178,7 +185,44 @@ class ManagerProvider extends Component<Props, State> {
       }
     });
 
-    api.on(SET_STORIES, (data: { stories: StoriesRaw }) => {
+    api.on(SET_STORIES, function handleSetStories(data: { stories: StoriesRaw }) {
+      const defaultMapper: Mapper = (b: InceptionRef, a: Story): Story => {
+        return { ...a, kind: `${b.id}/${a.kind.replace('|', '/')}` };
+      };
+
+      const { source } = this;
+      const { refs, mapper = defaultMapper } = provider.getConfig();
+
+      const { origin, pathname } = location;
+
+      if (refs) {
+        const refsList = Object.entries(refs);
+        const match = source === origin || source === `${origin + pathname}iframe.html`;
+
+        if (!match) {
+          const [id, u] =
+            refsList.find(([, url]: any) => url === source) ||
+            refsList.find(([, url]: any) => url.match(source));
+
+          Object.entries(data.stories).forEach(([k, v]) => {
+            // eslint-disable-next-line no-param-reassign
+            delete data.stories[k];
+
+            const mapped = mapper ? mapper({ id, url: u }, v) : v;
+
+            if (mapped) {
+              // eslint-disable-next-line no-param-reassign
+              data.stories[`${id}_${mapped.id}`] = {
+                ...mapped,
+                id: `${id}_${mapped.id}`,
+                knownAs: k,
+                source: u,
+              };
+            }
+          });
+        }
+      }
```

Comment:
I find the code above super dense and it's kinda hard to read and grok what its doing. Can we break it up into a few named methods, add comments  etc. to make it more obvious what the logic is?

Reply:
I've tried to make it simpler, WDYT?